### PR TITLE
issue: 933410 Support IPoIB for CX4

### PR DESCRIPTION
**Pull request by igor-ivanov**
These changes add IPoIB functionality for CX4.
This capability can be used with related OFED.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1245/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/869/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1248/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/870/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1260/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1276/ for details (Mellanox internal link).

File: src/vma/dev/qp_mgr.cpp [hunk 1]
```diff
@@ -797,4 +819,20 @@ void qp_mgr_ib::update_pkey_index()
...
+	 * decision can be made just after attempt to create qp. The value of
+	 * m_underly_qpn is reverted to zero if function to qp creation returns
+	 * failure.
+	 * So zero value for this field means no such capability.
+	 * Note: mlx4 does not support this capability. Disable it explicitly because dynamic check
+	 * using ibv_create_qp does not help
+	 */
+	if (strncmp(m_p_ib_ctx_handler->get_ibv_device()->name, "mlx4", 4)) {
+		m_underly_qpn = m_p_ring->get_qpn();
+	}
```
Comment (OphirMunk) [inline src/vma/dev/qp_mgr.cpp]: Prefer an explicit assignment:
```
} else
        m_underly_qpn = 0;
}

File: src/vma/util/verbs_extra.h [hunk 2]
```diff
@@ -171,7 +176,13 @@ typedef struct ibv_flow_spec_ib			vma_ibv_flow_spec_ib;
 typedef struct ibv_flow_spec_eth		vma_ibv_flow_spec_eth;
 typedef struct ibv_flow_spec_ipv4		vma_ibv_flow_spec_ipv4;
 typedef struct ibv_flow_spec_tcp_udp		vma_ibv_flow_spec_tcp_udp;
+
 #else //new MLNX_OFED verbs (2.2 and newer)
+
+#define vma_ibv_create_qp(pd, attr)             ibv_exp_create_qp((pd)->context, attr)
+typedef struct ibv_exp_qp_init_attr             vma_ibv_qp_init_attr;
+#define vma_ibv_qp_init_attr_comp_mask(_pd, _attr)	\
+	{ (_attr).pd = _pd; (_attr).comp_mask = IBV_EXP_QP_INIT_ATTR_PD; }
```
Comment (igor-ivanov) [inline src/vma/util/verbs_extra.h]: it can be done

File: src/vma/dev/qp_mgr.cpp [hunk 3]
```diff
@@ -759,26 +761,46 @@ void qp_mgr_ib::modify_qp_to_ready_state()
...
 
-int qp_mgr_ib::prepare_ibv_qp(struct ibv_qp_init_attr& qp_init_attr)
+int qp_mgr_ib::prepare_ibv_qp(vma_ibv_qp_init_attr& qp_init_attr)
 {
 	qp_logdbg("");
 	int ret = 0;
+
 	qp_init_attr.qp_type = IBV_QPT_UD;
-	m_qp = ibv_create_qp(m_p_ib_ctx_handler->get_ibv_pd(), &qp_init_attr);
+	vma_ibv_qp_init_attr_comp_mask(m_p_ib_ctx_handler->get_ibv_pd(), qp_init_attr);
```
Comment (igor-ivanov) [inline src/vma/dev/qp_mgr.cpp]: Sorry, I do not understand your comment.

File: src/vma/dev/qp_mgr.cpp [hunk 4]
```diff
@@ -759,26 +761,46 @@ void qp_mgr_ib::modify_qp_to_ready_state()
...
+	}
+#endif /* DEFINED_IBV_EXP_QP_INIT_ATTR_ASSOCIATED_QPN */
+
+	m_qp = vma_ibv_create_qp(m_p_ib_ctx_handler->get_ibv_pd(), &qp_init_attr);
+#ifdef DEFINED_IBV_EXP_QP_INIT_ATTR_ASSOCIATED_QPN
+	if (!m_qp && m_underly_qpn) {
+		qp_logdbg("ibv_create_qp failed to use underly qpn (errno=%d %m)", errno);
+
+		/* try to use traditional approach */
+		qp_init_attr.comp_mask &= ~IBV_EXP_QP_INIT_ATTR_ASSOCIATED_QPN;
```
Comment (igor-ivanov) [inline src/vma/dev/qp_mgr.cpp]: If we use experimental verbs we must set comp_mask with IBV_EXP_QP_INIT_ATTR_PD

File: src/vma/dev/qp_mgr.cpp (hunk 1 above)
Comment (igor-ivanov) [inline src/vma/dev/qp_mgr.cpp]: it is done explicitly in class constructor

**Comment by OphirMunk** [comment]
Merged #314

**Pull request closed without merging**